Mypy plugin crashes on an assertion error

### Describe the problem.

When upgrading the extensions package to `proxystore==0.6.5`, I got the following mypy crash.

```
proxystore-extensions$ mypy proxystore_ex/plugins/distributed.py --show
-traceback
proxystore_ex/plugins/distributed.py:406: error: INTERNAL ERROR -- Please try using mypy master on GitHub:
https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 1.10.0
Traceback (most recent call last):
  File "mypy/checkexpr.py", line 5760, in accept
  File "mypy/nodes.py", line 1831, in accept
  File "mypy/checkexpr.py", line 3243, in visit_member_expr
  File "mypy/checkexpr.py", line 3264, in analyze_ordinary_member_access
  File "mypy/checkmember.py", line 205, in analyze_member_access
  File "mypy/checkmember.py", line 229, in _analyze_member_access
  File "mypy/checkmember.py", line 466, in analyze_union_member_access
  File "mypy/checkmember.py", line 224, in _analyze_member_access
  File "mypy/checkmember.py", line 352, in analyze_instance_member_access
  File "mypy/checkmember.py", line 573, in analyze_member_var_access
  File "/home/jgpaul/workspace/proxystore-extensions/venv/lib/python3.11/site-packages/proxystore/mypy_plugin.py", line 90, in proxy_attribute_access
    assert isinstance(ctx.type, Instance)
AssertionError:
proxystore_ex/plugins/distributed.py:406: : note: use --pdb to drop into pdb
```

The crash happens on this line: https://github.com/proxystore/extensions/blob/b0aba0fc840316e8f1a716343194d7d4d3ad2359/proxystore_ex/plugins/distributed.py#L406

In v0.6.5, `__factory__` no longer exists so I'd expect a mypy error but not a crash. Fixing those lines to use `get_factory()` fixes the crash, but I'd like to avoid crashing entirely.

Here's a minimal reproducer:
```python
from typing import Callable, TypeVar
from proxystore.proxy import Proxy

T = TypeVar('T')

def foo(x: T | Proxy[T]) -> T | Callable[[], T]:
    if isinstance(x, Proxy):
        return x.__factory__  # crashes
    else:
        return x

x = Proxy(lambda: 42)
x.__factory__  # errors correctly that int has no attribute __factory__
```
The error seems specific to `foo()`'s parameter `x` being `T | Proxy[T]`. Removing the union and just having `x: Proxy[T]` makes the crash go away.

### How did you install ProxyStore?

```console
$ pip install proxystore
```


### ProxyStore Version

v0.6.5

### Python Version

All

### OS and Platform

All

--- a/proxystore/mypy_plugin.py
+++ b/proxystore/mypy_plugin.py
@@ -53,16 +53,29 @@ in your
 
 from __future__ import annotations
 
+import functools
+import sys
 from typing import Callable
+from typing import TypeVar
+
+if sys.version_info >= (3, 10):  # pragma: >=3.10 cover
+    from typing import ParamSpec
+else:  # pragma: <3.10 cover
+    from typing_extensions import ParamSpec
 
 from mypy.checkmember import analyze_member_access
 from mypy.options import Options
 from mypy.plugin import AttributeContext
 from mypy.plugin import Plugin
+from mypy.types import AnyType
 from mypy.types import get_proper_type
 from mypy.types import Instance
 from mypy.types import Type
+from mypy.types import TypeOfAny
 from mypy.types import TypeVarType
+
+P = ParamSpec('P')
+T = TypeVar('T')
 
 PROXY_FULLNAME = 'proxystore.proxy.Proxy'
 
@@ -83,6 +96,22 @@ class ProxyStoreMypyPlugin(Plugin):  # noqa: D101
         return None
 
 
+def _assertion_fallback(function: Callable[P, Type]) -> Callable[P, Type]:
+    # Decorator which catches AssertionErrors and returns AnyType
+    # to indicate that the plugin does not know how to handle that case
+    # and will default back to Any.
+    # https://github.com/dry-python/returns/blob/dda187d78fe405d7d1234ffaffc99d8264f854dc/returns/contrib/mypy/_typeops/fallback.py
+    @functools.wraps(function)
+    def decorator(*args: P.args, **kwargs: P.kwargs) -> Type:
+        try:
+            return function(*args, **kwargs)
+        except AssertionError:
+            return AnyType(TypeOfAny.implementation_artifact)
+
+    return decorator
+
+
+@_assertion_fallback
 def proxy_attribute_access(ctx: AttributeContext) -> Type:  # noqa: D103
     # These assertions should be covered by
     # ProxyStoreMyPyPlugin.get_attribute_hook, the only function which

--- a/tests/mypy_plugin_test.py
+++ b/tests/mypy_plugin_test.py
@@ -138,3 +138,19 @@ def test_proxy_bad_attribute_use() -> None:
 
     # Will fail because mypy should have inferred x is an int.
     expects_str(x)  # type: ignore[arg-type]
+
+
+def test_union_type_bad_attribute_crash() -> None:
+    # https://github.com/proxystore/proxystore/issues/559
+    def foo(x: T | Proxy[T]) -> T | Callable[[], T]:  # pragma: no cover
+        # Note: this function should not actually be called! It will error!
+        if isinstance(x, Proxy):
+            return x.__factory__  # In #559, mypy crashes on this line
+        else:
+            return x
+
+    x = Proxy(lambda: 42)
+    try:
+        _ = x.__factory__  # type: ignore[attr-defined]
+    except AttributeError:
+        pass

